unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,432 @@
+        }
+        let bytes = self.block_buffer.len();
+
+        let inner = FlatUnixFs {
+            links: Vec::new(),
+            data: UnixFs {
+                Type: UnixFsType::File,
+                Data: Some(Cow::Borrowed(self.block_buffer.as_slice())),
+                filesize: Some(self.block_buffer.len() as u64),
+                // no blocksizes as there are no links
+                ..Default::default()
+            },
+        };
+
+        let (cid, vec) = render_and_hash(inner);
+
+        let total_size = vec.len();
+
+        self.block_buffer.clear();
+
+        // leafs always go to the lowest level
+        if self.unflushed_links.is_empty() {
+            self.unflushed_links.push(Vec::new());
+        }
+
+        self.unflushed_links[0].push((cid.clone(), total_size as u64, bytes as u64));
+
+        Some((cid, vec))
+    }
+
+    // FIXME: collapse the min_links and all into single type to avoid boolean args
+    fn flush_buffered_links(&mut self, min_links: NonZeroUsize, all: bool) -> Vec<(Cid, Vec<u8>)> {
+        /*
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|\   /
+        links#2 |-------------------------------|                         ^^^
+                                                                    one short
+
+        #flush_buffered_links(...) first iterations:
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|==1==|
+        links#2 |-------------------------------|==========================2==|
+
+        new blocks #1 and #2
+
+        #flush_buffered_links(...) last iteration:
+
+        file    |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -|
+        links#0 |-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|-|
+        links#1 |-------|-------|-------|-------|-------|-------|-------|--1--|
+        links#2 |-------------------------------|--------------------------2--|
+        links#3 |==========================================================3==|
+
+        new block #3 (the root block)
+        */
+        let mut ret = Vec::new();
+
+        for level in 0.. {
+            if !all {
+                if self.unflushed_links[level].len() < min_links.get() {
+                    break;
+                }
+            } else if self.unflushed_links[level].len() == 1 {
+                // TODO: combine with above?
+                // we need to break here as otherwise we'd be looping for ever
+                break;
+            }
+
+            let mut links = Vec::with_capacity(self.unflushed_links.len());
+            let mut blocksizes = Vec::with_capacity(self.unflushed_links.len());
```

Comment:
**if** these 2 objects are always the same size (and especially if they're also traversed together), it would be cool if we could allocate them together, i.e. have a `Vec<(link, blocksize)>`

Reply:
They need to be put into the FlatUnixFs object on lines 213..214 but ... they could be taken out of there and reused!

As for the partitioning trick we need to do on lines 198..207  I can't see a non-forloopy answer.

Reply:
This algorithm is broken for balanced ... need to think this over.

Reply:
Rewrote the algoritm here, now we have perhaps different perf issues with the repeat linear search but ... I think it's quite negleable. I'll add a comment there. I added these reuses.